// @addo/addo-printer/lib/autoreplyprint.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.autoreplyprint = void 0;
const ffi_1 = __importDefault(require("ffi"));
const ref_1 = __importDefault(require("ref"));
const fs_1 = __importDefault(require("fs"));
const libPath = '/usr/local/lib/libautoreplyprint.so';
if (!fs_1.default.existsSync(libPath)) {
    console.error(`Printer library seems not to be installed.\nPlease make sure it is located in ${libPath} and configured correctly with ldconfig command.\n\n`);
}
const types = ref_1.default.types;
const refType = ref_1.default.refType;
const func_t = 'pointer';
const char_t = types.char;
const uint_t = types.uint;
const int_t = types.int;
const void_t = types.void;
const ushort_t = types.ushort;
const uchar_t = types.uchar;
const longlong_t = types.longlong;
const double_t = types.double;
const wchar_t = types.int16;
const char_p = refType(char_t);
const uint_p = refType(uint_t);
const int_p = refType(int_t);
const void_p = refType(void_t);
const uchar_p = refType(uchar_t);
const longlong_p = refType(longlong_t);
const double_p = refType(double_t);
const wchar_p = refType(wchar_t);
const string_t = types.CString;
exports.autoreplyprint = ffi_1.default.Library(libPath, {
    CP_Library_Version: [char_p, []],
    CP_Port_EnumCom: [uint_t, [char_p, uint_t, uint_p]],
    CP_Port_EnumLpt: [uint_t, [char_p, uint_t, uint_p]],
    CP_Port_EnumUsb: [uint_t, [char_p, uint_t, uint_p]],
    CP_Port_EnumNetPrinter: [void_t, [uint_t, int_p, func_t, void_p]],
    CP_Port_EnumBtDevice: [void_t, [uint_t, int_p, func_t, void_p]],
    CP_Port_EnumBleDevice: [void_t, [uint_t, int_p, func_t, void_p]],
    CP_Port_OpenCom: [void_p, [string_t, uint_t, uint_t, uint_t, uint_t, uint_t, int_t]],
    CP_Port_OpenLpt: [void_p, [string_t]],
    CP_Port_OpenUsb: [void_p, [string_t, int_t]],
    CP_Port_OpenTcp: [void_p, [string_t, string_t, ushort_t, uint_t, int_t]],
    CP_Port_OpenBtSpp: [void_p, [string_t, int_t]],
    CP_Port_OpenBtBle: [void_p, [string_t, int_t]],
    CP_Port_Write: [int_t, [void_p, uchar_p, uint_t, uint_t]],
    CP_Port_Read: [int_t, [void_p, uchar_p, uint_t, uint_t]],
    CP_Port_ReadUntilByte: [int_t, [void_p, uchar_p, uint_t, uint_t, uchar_t]],
    CP_Port_Available: [int_t, [void_p]],
    CP_Port_SkipAvailable: [int_t, [void_p]],
    CP_Port_IsConnectionValid: [int_t, [void_p]],
    CP_Port_IsOpened: [int_t, [void_p]],
    CP_Port_Close: [int_t, [void_p]],
    CP_Port_AddOnPortOpenedEvent: [int_t, [func_t, void_p]],
    CP_Port_AddOnPortOpenFailedEvent: [int_t, [func_t, void_p]],
    CP_Port_AddOnPortClosedEvent: [int_t, [func_t, void_p]],
    CP_Port_RemoveOnPortOpenedEvent: [int_t, [func_t]],
    CP_Port_RemoveOnPortOpenFailedEvent: [int_t, [func_t]],
    CP_Port_RemoveOnPortClosedEvent: [int_t, [func_t]],
    CP_Printer_AddOnPrinterStatusEvent: [int_t, [func_t, void_p]],
    CP_Printer_AddOnPrinterReceivedEvent: [int_t, [func_t, void_p]],
    CP_Printer_AddOnPrinterPrintedEvent: [int_t, [func_t, void_p]],
    CP_Printer_RemoveOnPrinterStatusEvent: [int_t, [func_t]],
    CP_Printer_RemoveOnPrinterReceivedEvent: [int_t, [func_t]],
    CP_Printer_RemoveOnPrinterPrintedEvent: [int_t, [func_t]],
    CP_Printer_GetPrinterResolutionInfo: [int_t, [void_p, uint_p, uint_p, uint_p]],
    CP_Printer_GetPrinterFirmwareVersion: [int_t, [void_p, char_p, uint_t, uint_p]],
    CP_Printer_GetPrinterStatusInfo: [int_t, [void_p, longlong_p, longlong_p, longlong_p]],
    CP_Printer_GetPrinterReceivedInfo: [int_t, [void_p, uint_p, longlong_p]],
    CP_Printer_GetPrinterPrintedInfo: [int_t, [void_p, uint_p, longlong_p]],
    CP_Printer_GetPrinterLabelPositionAdjustmentInfo: [int_t, [void_p, double_p, double_p, longlong_p]],
    CP_Printer_SetPrinterLabelPositionAdjustmentInfo: [int_t, [void_p, double_t, double_t]],
    CP_Printer_ClearPrinterBuffer: [int_t, [void_p]],
    CP_Printer_ClearPrinterError: [int_t, [void_p]],
    CP_Pos_QueryRTStatus: [int_t, [void_p, uint_t]],
    CP_Pos_QueryPrintResult: [int_t, [void_p, uint_t, uint_t]],
    CP_Pos_KickOutDrawer: [int_t, [void_p, int_t, int_t, int_t]],
    CP_Pos_Beep: [int_t, [void_p, int_t, int_t]],
    CP_Pos_FeedAndHalfCutPaper: [int_t, [void_p]],
    CP_Pos_FullCutPaper: [int_t, [void_p]],
    CP_Pos_HalfCutPaper: [int_t, [void_p]],
    CP_Pos_FeedLine: [int_t, [void_p, int_t]],
    CP_Pos_FeedDot: [int_t, [void_p, int_t]],
    CP_Pos_PrintSelfTestPage: [int_t, [void_p]],
    CP_Pos_PrintText: [int_t, [void_p, string_t]],
    CP_Pos_PrintTextInUTF8: [int_t, [void_p, wchar_p]],
    CP_Pos_PrintTextInGBK: [int_t, [void_p, wchar_p]],
    CP_Pos_PrintTextInBIG5: [int_t, [void_p, wchar_p]],
    CP_Pos_PrintTextInShiftJIS: [int_t, [void_p, wchar_p]],
    CP_Pos_PrintTextInEUCKR: [int_t, [void_p, wchar_p]],
    CP_Pos_PrintTextInBytes: [int_t, [void_p, char_p, int_t]],
    CP_Pos_PrintBarcode: [int_t, [void_p, uint_t, string_t]],
    CP_Pos_PrintQRCode: [int_t, [void_p, int_t, uint_t, string_t]],
    CP_Pos_PrintQRCodeUseEpsonCmd: [int_t, [void_p, int_t, uint_t, string_t]],
    CP_Pos_PrintDoubleQRCode: [int_t, [void_p, int_t, int_t, int_t, uint_t, string_t, int_t, int_t, uint_t, string_t]],
    CP_Pos_PrintPDF417BarcodeUseEpsonCmd: [int_t, [void_p, int_t, int_t, int_t, int_t, int_t, int_t, string_t]],
    CP_Pos_PrintRasterImageFromFile: [int_t, [void_p, int_t, int_t, string_t, uint_t, uint_t]],
    CP_Pos_PrintRasterImageFromData: [int_t, [void_p, int_t, int_t, uchar_p, uint_t, uint_t, uint_t]],
    CP_Pos_PrintRasterImageFromPixels: [int_t, [void_p, uchar_p, uint_t, int_t, int_t, int_t, uint_t, uint_t, uint_t]],
    CP_Pos_PrintHorizontalLine: [int_t, [void_p, int_t, int_t]],
    CP_Pos_PrintHorizontalLineSpecifyThickness: [int_t, [void_p, int_t, int_t, int_t]],
    CP_Pos_PrintMultipleHorizontalLinesAtOneRow: [int_t, [void_p, int_t, int_p, int_p]],
    CP_Pos_ResetPrinter: [int_t, [void_p]],
    CP_Pos_SetPrintSpeed: [int_t, [void_p, int_t]],
    CP_Pos_SetPrintDensity: [int_t, [void_p, int_t]],
    CP_Pos_SetSingleByteMode: [int_t, [void_p]],
    CP_Pos_SetCharacterSet: [int_t, [void_p, uint_t]],
    CP_Pos_SetCharacterCodepage: [int_t, [void_p, uint_t]],
    CP_Pos_SetMultiByteMode: [int_t, [void_p]],
    CP_Pos_SetMultiByteEncoding: [int_t, [void_p, uint_t]],
    CP_Pos_SetMovementUnit: [int_t, [void_p, int_t, int_t]],
    CP_Pos_SetPrintAreaLeftMargin: [int_t, [void_p, int_t]],
    CP_Pos_SetPrintAreaWidth: [int_t, [void_p, int_t]],
    CP_Pos_SetHorizontalAbsolutePrintPosition: [int_t, [void_p, int_t]],
    CP_Pos_SetHorizontalRelativePrintPosition: [int_t, [void_p, int_t]],
    CP_Pos_SetVerticalAbsolutePrintPosition: [int_t, [void_p, int_t]],
    CP_Pos_SetVerticalRelativePrintPosition: [int_t, [void_p, int_t]],
    CP_Pos_SetAlignment: [int_t, [void_p, uint_t]],
    CP_Pos_SetTextScale: [int_t, [void_p, int_t, int_t]],
    CP_Pos_SetAsciiTextFontType: [int_t, [void_p, int_t]],
    CP_Pos_SetTextBold: [int_t, [void_p, int_t]],
    CP_Pos_SetTextUnderline: [int_t, [void_p, int_t]],
    CP_Pos_SetTextUpsideDown: [int_t, [void_p, int_t]],
    CP_Pos_SetTextWhiteOnBlack: [int_t, [void_p, int_t]],
    CP_Pos_SetTextRotate: [int_t, [void_p, int_t]],
    CP_Pos_SetTextLineHeight: [int_t, [void_p, int_t]],
    CP_Pos_SetAsciiTextCharRightSpacing: [int_t, [void_p, int_t]],
    CP_Pos_SetKanjiTextCharSpacing: [int_t, [void_p, int_t, int_t]],
    CP_Pos_SetBarcodeUnitWidth: [int_t, [void_p, int_t]],
    CP_Pos_SetBarcodeHeight: [int_t, [void_p, int_t]],
    CP_Pos_SetBarcodeReadableTextFontType: [int_t, [void_p, int_t]],
    CP_Pos_SetBarcodeReadableTextPosition: [int_t, [void_p, uint_t]],
    CP_Page_SelectPageMode: [int_t, [void_p]],
    CP_Page_SelectPageModeEx: [int_t, [void_p, int_t, int_t, int_t, int_t, int_t, int_t]],
    CP_Page_ExitPageMode: [int_t, [void_p]],
    CP_Page_PrintPage: [int_t, [void_p]],
    CP_Page_ClearPage: [int_t, [void_p]],
    CP_Page_SetPageArea: [int_t, [void_p, int_t, int_t, int_t, int_t]],
    CP_Page_SetPageDrawDirection: [int_t, [void_p, uint_t]],
    CP_Page_DrawRect: [int_t, [void_p, int_t, int_t, int_t, int_t, int_t]],
    CP_Page_DrawBox: [int_t, [void_p, int_t, int_t, int_t, int_t, int_t, int_t]],
    CP_Page_DrawText: [int_t, [void_p, int_t, int_t, string_t]],
    CP_Page_DrawTextInUTF8: [int_t, [void_p, int_t, int_t, wchar_p]],
    CP_Page_DrawTextInGBK: [int_t, [void_p, int_t, int_t, wchar_p]],
    CP_Page_DrawTextInBIG5: [int_t, [void_p, int_t, int_t, wchar_p]],
    CP_Page_DrawTextInShiftJIS: [int_t, [void_p, int_t, int_t, wchar_p]],
    CP_Page_DrawTextInEUCKR: [int_t, [void_p, int_t, int_t, wchar_p]],
    CP_Page_DrawBarcode: [int_t, [void_p, int_t, int_t, uint_t, string_t]],
    CP_Page_DrawQRCode: [int_t, [void_p, int_t, int_t, int_t, uint_t, string_t]],
    CP_Page_DrawRasterImageFromFile: [int_t, [void_p, int_t, int_t, int_t, int_t, string_t, uint_t]],
    CP_Page_DrawRasterImageFromData: [int_t, [void_p, int_t, int_t, int_t, int_t, uchar_p, uint_t, uint_t]],
    CP_Page_DrawRasterImageFromPixels: [int_t, [void_p, int_t, int_t, uchar_p, uint_t, int_t, int_t, int_t, uint_t, uint_t]],
    CP_BlackMark_EnableBlackMarkMode: [int_t, [void_p]],
    CP_BlackMark_DisableBlackMarkMode: [int_t, [void_p]],
    CP_BlackMark_SetBlackMarkMaxFindLength: [int_t, [void_p, int_t]],
    CP_BlackMark_FindNextBlackMark: [int_t, [void_p]],
    CP_BlackMark_SetBlackMarkPaperPrintPosition: [int_t, [void_p, int_t]],
    CP_BlackMark_SetBlackMarkPaperCutPosition: [int_t, [void_p, int_t]],
    CP_BlackMark_FullCutBlackMarkPaper: [int_t, [void_p]],
    CP_BlackMark_HalfCutBlackMarkPaper: [int_t, [void_p]],
    CP_Label_EnableLabelMode: [int_t, [void_p]],
    CP_Label_DisableLabelMode: [int_t, [void_p]],
    CP_Label_CalibrateLabel: [int_t, [void_p]],
    CP_Label_FeedLabel: [int_t, [void_p]],
    CP_Label_PageBegin: [int_t, [void_p, int_t, int_t, int_t, int_t, uint_t]],
    CP_Label_PagePrint: [int_t, [void_p, int_t]],
    CP_Label_DrawText: [int_t, [void_p, int_t, int_t, int_t, int_t, string_t]],
    CP_Label_DrawTextInUTF8: [int_t, [void_p, int_t, int_t, int_t, int_t, wchar_p]],
    CP_Label_DrawTextInGBK: [int_t, [void_p, int_t, int_t, int_t, int_t, wchar_p]],
    CP_Label_DrawTextInBIG5: [int_t, [void_p, int_t, int_t, int_t, int_t, wchar_p]],
    CP_Label_DrawTextInShiftJIS: [int_t, [void_p, int_t, int_t, int_t, int_t, wchar_p]],
    CP_Label_DrawTextInEUCKR: [int_t, [void_p, int_t, int_t, int_t, int_t, wchar_p]],
    CP_Label_DrawTextInBytes: [int_t, [void_p, int_t, int_t, int_t, int_t, char_p, int_t]],
    CP_Label_DrawBarcode: [int_t, [void_p, int_t, int_t, uint_t, uint_t, int_t, int_t, uint_t, string_t]],
    CP_Label_DrawQRCode: [int_t, [void_p, int_t, int_t, int_t, uint_t, int_t, uint_t, string_t]],
    CP_Label_DrawQRCodeInUTF8: [int_t, [void_p, int_t, int_t, int_t, uint_t, int_t, uint_t, wchar_p]],
    CP_Label_DrawQRCodeInBytes: [int_t, [void_p, int_t, int_t, int_t, uint_t, int_t, uint_t, char_p, int_t]],
    CP_Label_DrawPDF417Code: [int_t, [void_p, int_t, int_t, int_t, int_t, int_t, int_t, uint_t, string_t]],
    CP_Label_DrawImageFromFile: [int_t, [void_p, int_t, int_t, int_t, int_t, string_t, uint_t]],
    CP_Label_DrawImageFromData: [int_t, [void_p, int_t, int_t, int_t, int_t, uchar_p, uint_t, uint_t]],
    CP_Label_DrawImageFromPixels: [int_t, [void_p, int_t, int_t, uchar_p, uint_t, int_t, int_t, int_t, uint_t, uint_t]],
    CP_Label_DrawLine: [int_t, [void_p, int_t, int_t, int_t, int_t, int_t, uint_t]],
    CP_Label_DrawRect: [int_t, [void_p, int_t, int_t, int_t, int_t, uint_t]],
    CP_Label_DrawBox: [int_t, [void_p, int_t, int_t, int_t, int_t, int_t, uint_t]]
});
//export default autoreplyprint;
//module.exports = autoreplyprint;
