// @addo/addo-printer/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ref_1 = require("ref");
const autoreplyprint_1 = require("./autoreplyprint");
const printer_alignment_1 = require("./core/printer-alignment");
const printer_status_1 = require("./core/printer-status");
const rxjs_1 = require("rxjs");
const DEFAULT_AUTOREPLY_MODE = 0;
class Printer {
    constructor(dev, mode = DEFAULT_AUTOREPLY_MODE) {
        const buf = ref_1.allocCString('Printer opened');
        this.handle = autoreplyprint_1.autoreplyprint.CP_Port_OpenUsb(dev, mode);
        console.log("Handle: ", this.handle);
    }
    printText(text) {
        return autoreplyprint_1.autoreplyprint.CP_Pos_PrintText(this.handle, text);
    }
    feedLine(numLines) {
        return autoreplyprint_1.autoreplyprint.CP_Pos_FeedLine(this.handle, numLines);
    }
    getPrinterStatus(timeout) {
        return autoreplyprint_1.autoreplyprint.CP_Pos_QueryRTStatus(this.handle, timeout);
    }
    cutPaper(fullCut = false) {
        if (fullCut)
            return autoreplyprint_1.autoreplyprint.CP_Pos_FullCutPaper(this.handle);
        else
            return autoreplyprint_1.autoreplyprint.CP_Pos_HalfCutPaper(this.handle);
    }
    feedAndHalfCutPaper() {
        return autoreplyprint_1.autoreplyprint.CP_Pos_FeedAndHalfCutPaper(this.handle);
    }
    close() {
        return autoreplyprint_1.autoreplyprint.CP_Port_Close(this.handle);
    }
    reset() {
        return autoreplyprint_1.autoreplyprint.CP_Pos_ResetPrinter(this.handle);
    }
    clear() {
        return autoreplyprint_1.autoreplyprint.CP_Printer_ClearPrinterBuffer(this.handle)
            && autoreplyprint_1.autoreplyprint.CP_Printer_ClearPrinterError(this.handle)
            && autoreplyprint_1.autoreplyprint.CP_Pos_ResetPrinter(this.handle);
    }
    version() {
        return autoreplyprint_1.autoreplyprint.CP_Library_Version();
    }
    printTestPage() {
        return autoreplyprint_1.autoreplyprint.CP_Pos_PrintSelfTestPage(this.handle);
    }
    /**
     * Set print alignment
     *
     * @param nAlignment print alignment, value are defined as follow:
     *      - 0 -> align left
     *      - 1 ->  align center
     *      - 2 ->  align right
     */
    setPrintAlignment(nAlignment) {
        return autoreplyprint_1.autoreplyprint.CP_Pos_SetAlignment(this.handle, nAlignment);
    }
    printTicket(ticket, clear = false) {
        return new Promise((resolve, reject) => {
            let status = [];
            status.push(autoreplyprint_1.autoreplyprint.CP_Pos_SetMultiByteEncoding(this.handle, 1));
            status.push(this.setPrintAlignment(printer_alignment_1.PrinterAlignment.center));
            status.push(this.printText(ticket.date));
            status.push(this.feedLine(1));
            status.push(autoreplyprint_1.autoreplyprint.CP_Pos_SetTextScale(this.handle, 1, 1));
            status.push(this.printText(ticket.location));
            status.push(autoreplyprint_1.autoreplyprint.CP_Pos_SetTextScale(this.handle, 0, 0));
            status.push(this.feedLine(1));
            status.push(autoreplyprint_1.autoreplyprint.CP_Pos_SetTextScale(this.handle, 1, 1));
            status.push(this.printText(`${ticket.product_name} ${ticket.price}${ticket.currency}`));
            status.push(this.feedLine(2));
            status.push(autoreplyprint_1.autoreplyprint.CP_Pos_SetTextScale(this.handle, 0, 0));
            status.push(autoreplyprint_1.autoreplyprint.CP_Pos_SetTextRotate(this.handle, ticket.rotateWatermark ? 1 : 0));
            status.push(this.printText(ticket.watermark));
            status.push(this.feedLine(1));
            status.push(autoreplyprint_1.autoreplyprint.CP_Pos_SetTextRotate(this.handle, 0));
            status.push(this.printText(ticket.printed));
            status.push(this.feedLine(2));
            status.push(this.cutPaper());
            if (clear) {
                status.push(autoreplyprint_1.autoreplyprint.CP_Printer_ClearPrinterBuffer(this.handle));
                status.push(this.clear());
            }
            else {
                status.push('-');
                status.push('-');
            }
            let result = { status: status.join(''), code: "_NO_PRINTER_CHECK", };
            // if (lib.CP_Pos_QueryPrintResult(this.handle, 1, 30000)) {
            //     result = this.getPrinterInfo()
            //     console.log(result);
            // }
            resolve(result);
        });
    }
    printTestTicket() {
        console.log(autoreplyprint_1.autoreplyprint.CP_Printer_ClearPrinterBuffer(this.handle), this.clear(), autoreplyprint_1.autoreplyprint.CP_Pos_SetMultiByteEncoding(this.handle, 1), this.setPrintAlignment(printer_alignment_1.PrinterAlignment.center), this.printText("Sab 14 Ago"), this.feedLine(1), autoreplyprint_1.autoreplyprint.CP_Pos_SetTextScale(this.handle, 1, 1), this.printText("Cocoricò"), autoreplyprint_1.autoreplyprint.CP_Pos_SetTextScale(this.handle, 0, 0), this.feedLine(1), 
        //lib.CP_Pos_SetTextBold(this.handle, 1),
        autoreplyprint_1.autoreplyprint.CP_Pos_SetTextScale(this.handle, 1, 1), 
        // this.printText("DRINK"),
        // this.feedLine(1),
        // this.printText("5.00 €"),
        this.printText("ANALCOLICO 15€"), this.feedLine(2), autoreplyprint_1.autoreplyprint.CP_Pos_SetTextScale(this.handle, 0, 0), autoreplyprint_1.autoreplyprint.CP_Pos_SetTextRotate(this.handle, 1), this.printText("¤-¤-¤-¤-¤-¤-¤-¤"), this.feedLine(1), autoreplyprint_1.autoreplyprint.CP_Pos_SetTextRotate(this.handle, 0), this.printText("Validato il 21/10/2020"), this.feedLine(2), this.cutPaper(), this.clear(), this.close());
    }
    getPrinterInfo() {
        let result = { status: 'ok' };
        let printer_error_status = 0;
        let printer_info_status = 0;
        let printer_received_byte_count = 0;
        let printer_printed_page_id = 0;
        let timestamp_ms_printer_status = 0;
        let timestamp_ms_printer_received = 0;
        let timestamp_ms_printer_printed = 0;
        if (autoreplyprint_1.autoreplyprint.CP_Printer_GetPrinterStatusInfo(this.handle, printer_error_status, printer_info_status, timestamp_ms_printer_status) &&
            autoreplyprint_1.autoreplyprint.CP_Printer_GetPrinterReceivedInfo(this.handle, printer_received_byte_count, timestamp_ms_printer_received) &&
            autoreplyprint_1.autoreplyprint.CP_Printer_GetPrinterPrintedInfo(this.handle, printer_printed_page_id, timestamp_ms_printer_printed)) {
            //char msg[1000] = { 0, };
            // printer_error_status
            if (printer_error_status) {
                result.status = 'ko';
                result.printer_errors = printer_status_1.PrinterStatus.getStatus(printer_error_status);
            }
            // printer_info_status
            // {
            //     if (CP_PRINTERSTATUS_INFO_LABELMODE(printer_info_status))
            //         sprintf(& msg[strlen(msg)], "%s", "[Label Mode]");
            //     if (CP_PRINTERSTATUS_INFO_LABELPAPER(printer_info_status))
            //         sprintf(& msg[strlen(msg)], "%s", "[Label Paper]");
            //     if (CP_PRINTERSTATUS_INFO_PAPERNOFETCH(printer_info_status))
            //         sprintf(& msg[strlen(msg)], "%s", "[Paper Not Fetch]");
            //     sprintf(& msg[strlen(msg)], "%s", "\r\n");
            // }
            return result;
        }
    }
    onStatusChange() {
        return new rxjs_1.Observable((observer) => {
            let status = this.getPrinterStatus(15000);
            let last_status = 0;
            let interval = setInterval(() => {
                status = this.getPrinterStatus(15000);
                if (last_status != status) {
                    last_status = status;
                    observer.next(printer_status_1.PrinterStatus.getStatus(status));
                }
            }, 100);
            return {
                unsubscribe() {
                    clearInterval(interval);
                }
            };
        });
    }
}
exports.default = Printer;
