// @addo/common-api/lib/src/auth/auth.service.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Auth = void 0;
const JwtHelper = __importStar(require("jsonwebtoken"));
require("reflect-metadata");
const tsyringe_1 = require("tsyringe");
const cryptoEngine_1 = require("../utils/cryptoEngine");
let Auth = class Auth {
    constructor() {
        this.secret_key = "MXuqsy6:zYWaNy~]?iu;8R1KGF+(Jr12";
        this.iv = "MXuqsy6:zYWaNy~]";
        this.crypto_engine = new cryptoEngine_1.CryptoEngine(this.secret_key, this.iv);
    }
    setSecrets(key, iv) {
        this.secret_key = key || this.secret_key;
        this.iv = iv || this.iv;
    }
    validateToken(token) {
        let payload = this.getPayload(token);
        let id = this.crypto_engine.decrypt(payload.checksum).split("#")[0];
        return !!JwtHelper.verify(token, this.secret_key) && id == payload.id;
    }
    createToken(payload) {
        let checksum = this.crypto_engine.encrypt(payload.id + "#" + (payload.roles || []).join(","));
        payload.checksum = checksum;
        return JwtHelper.sign(payload, this.secret_key, { expiresIn: "90 days" });
    }
    getPayload(token) {
        try {
            let payload = JwtHelper.decode(token);
            return payload;
        }
        catch (error) {
            return null;
        }
    }
    tokenAsRole(token, role) {
        let payload = this.getPayload(token);
        let checksum = this.crypto_engine.decrypt(payload.checksum);
        let roles = (checksum.split("#")[1] || "").split(",");
        let id = checksum.split("#")[0];
        if (typeof role == "string") {
            return roles.indexOf(role) > -1 && id == payload.id;
        }
        return !!roles.find((r) => role.indexOf(r) > -1) && id == payload.id;
    }
    encryptPassword(password) {
        return this.crypto_engine.encrypt(password);
    }
};
Auth = __decorate([
    tsyringe_1.singleton(),
    __metadata("design:paramtypes", [])
], Auth);
exports.Auth = Auth;
