// @addo/common-api/lib/src/utils/http/servers/express/express.server.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExpressServer = void 0;
const express_1 = __importDefault(require("express"));
const node_cron_1 = __importDefault(require("node-cron"));
const body_parser_1 = __importDefault(require("body-parser"));
class ExpressServer {
    /**
     * Create a Express server with all routes
     *
     * @example
     * import { Response } from "../../models/response";
     * import { ExpressServer } from "./express.server";
     * import { ExpressJsMicroservice, HttpMethod } from "./expressjs.microservice";
     *
     * let endpoint = new ExpressJsMicroservice(
     *    { method: HttpMethod.GET, route: '/hello-world' },
     *    (http) => Promise.resolve<Response<string>>(new Response<string>('Hello world'))
     * );
     * let endpoint1 = new ExpressJsMicroservice(
     *    { method: HttpMethod.GET, route: '/hello-galaxy' },
     *    (http) => Promise.resolve<Response<string>>(new Response<string>('Hello galaxy'))
     * );
     *
     * let server = new ExpressServer(endpoint, endpoint1);
     * server.addMicroservice(
     *    new ExpressJsMicroservice(
     *        { method: HttpMethod.GET, route: '/hello-universe' },
     *        (http) => Promise.resolve<Response<string>>(new Response<string>('Hello universe!'))
     *    )
     * );
     * server.start(8080);
     */
    constructor(...endpoints) {
        this.routes = [];
        this.cronJobs = [];
        console.log(`Creating Express app`);
        this.app = express_1.default();
        this.app.use(body_parser_1.default.urlencoded({ extended: true }));
        this.app.use(body_parser_1.default.json());
        console.log(`Created Express app`);
        for (let microservice of endpoints) {
            this.addMicroservice(microservice);
        }
    }
    getRoutes() {
        return this.routes;
    }
    start(port = null) {
        // let routes = []
        // this.app._router.stack.forEach((r) => {
        //     if (r.route && r.route.path) {
        //         routes.push(r.route);
        //     }
        // })
        this.app.get('/docs/routes', (req, res) => {
            res.status(200).json(this.routes);
        });
        console.log(`======== ROUTES ========`);
        for (let route of this.routes) {
            console.log(`[${route.method}] ${route.route}`);
        }
        console.log(`========================`);
        this.app.listen(port || process.env.PORT, () => console.log(`Express app listening on port ${port || process.env.PORT}!`));
    }
    addMicroservice(microservice) {
        console.log(`Add new route: [${microservice.config.method}] ${microservice.config.route}`);
        this.app[microservice.config.method](microservice.config.route, microservice.getHandler());
        this.routes.push(microservice.config);
    }
    addCronJob(cronExpr, func) {
        if (node_cron_1.default.validate(cronExpr)) {
            this.cronJobs.push(node_cron_1.default.schedule(cronExpr, func));
            return true;
        }
        console.warn(`CronJob with ${cronExpr} was not valid!`);
        return false;
    }
}
exports.ExpressServer = ExpressServer;
