// @addo/common-api/lib/src/utils/http/servers/express/expressjs.microservice.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpMethod = exports.EndpointConfig = exports.ExpressJsMicroservice = void 0;
const microservice_1 = require("../../microservice");
const response_1 = require("../../models/response");
const http_request_1 = require("../../models/http.request");
class ExpressJsMicroservice extends microservice_1.Microservice {
    constructor(config, handler) {
        super(handler);
        this.config = config;
    }
    parseRequest(request) {
        let req = new http_request_1.Request();
        req.queryStringParameters = request.query;
        req.pathParameters = request.params;
        req.headers = JSON.parse(JSON.stringify(request.headers));
        req.body = request.body;
        req.original = request;
        return new http_request_1.HttpRequest(req);
    }
    async getAHandler(request, response) {
        this.http = this.parseRequest(request);
        let result = await this.handle(this.http);
        return result;
    }
    getHandler() {
        return (request, response) => {
            this.getAHandler(request, response)
                .then(res => {
                for (let key in res.headers) {
                    response.setHeader(key, res.headers[key]);
                }
                response.status(res.statusCode).send(res.body);
            }).catch(err => {
                if (err instanceof response_1.Response) {
                    for (let key in err.headers) {
                        response.setHeader(key, err.headers[key]);
                    }
                    response.status(err.statusCode).send(err.body);
                    return;
                }
                response.status(500).send({ code: "_GENERIC_ERROR", err, message: err.message });
            });
        };
    }
}
exports.ExpressJsMicroservice = ExpressJsMicroservice;
class EndpointConfig {
    constructor(route, method) {
        this.route = route;
        this.method = method;
    }
}
exports.EndpointConfig = EndpointConfig;
var HttpMethod;
(function (HttpMethod) {
    HttpMethod["GET"] = "get";
    HttpMethod["POST"] = "post";
    HttpMethod["PUT"] = "put";
    HttpMethod["PATCH"] = "patch";
    HttpMethod["DELETE"] = "delete";
})(HttpMethod = exports.HttpMethod || (exports.HttpMethod = {}));
