// @addo/common-api/lib/src/utils/http/aws.microservice.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AWSMicroservice = void 0;
const microservice_1 = require("../../utils/http/microservice");
const http_request_1 = require("./models/http.request");
class AWSMicroservice extends microservice_1.Microservice {
    constructor(handler) {
        super(handler);
    }
    parseRequest(request) {
        let req = new http_request_1.Request();
        req.body = request.body;
        req.headers = request.headers;
        req.pathParameters = request.pathParameters;
        req.queryStringParameters = request.queryStringParameters;
        req.original = request;
        return new http_request_1.HttpRequest(req);
    }
    async getAHandler(event, context, callback) {
        this.context = context;
        this.callback = callback;
        context.callbackWaitsForEmptyEventLoop = false;
        this.http = this.parseRequest(event);
        return await this.handle(this.http);
    }
}
exports.AWSMicroservice = AWSMicroservice;
