// @addo/common-api/lib/src/utils/http/microservice.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Microservice = void 0;
class Microservice {
    constructor(handle) {
        handle.bind(this);
        this.handle = handle;
    }
    parseRequest(request) {
        throw new Error("Method not implemented.");
    }
    handle(http) {
        throw new Error("Must override this method!");
    }
    getHandler() {
        let self = this;
        return this.getAHandler.bind(self);
    }
    async getAHandler(...args) {
        this.http = this.parseRequest(args[0]);
        let response = await this.handle(this.http);
        return response;
    }
    generateResponse(body, statusCode = 200, headers = null) {
        let default_headers = {
            "Access-Control-Allow-Origin": "*",
            "access-control-allow-headers": "Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token,X-Amz-User-Agent,X-User-Location",
        };
        return {
            statusCode: statusCode || 200,
            headers: headers
                ? Object.assign(default_headers, headers)
                : default_headers,
            body,
        };
    }
}
exports.Microservice = Microservice;
