// @addo/common-api/lib/src/models/xml-element.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MappingElement = exports.XmlElement = void 0;
class XmlElement {
    constructor() {
        this.mapping = [];
    }
    getXml(invoice_xml) {
        for (let el of this.mapping) {
            invoice_xml = this.convertXml(invoice_xml, el);
        }
        return invoice_xml;
    }
    convertXml(invoice_xml, el) {
        if (!this[el.key])
            return invoice_xml;
        if (Array.isArray(this[el.key])) {
            for (let o of this[el.key]) {
                invoice_xml += `<${el.key}>`;
                invoice_xml = o.getXml(invoice_xml);
                invoice_xml += `</${el.key}>`;
            }
        }
        else if (typeof this[el.key] === "object" && this[el.key] != null) {
            invoice_xml += `<${el.key}>`;
            invoice_xml = this[el.key].getXml(invoice_xml);
            invoice_xml += `</${el.key}>`;
        }
        else {
            invoice_xml += `<${el.key}>${el.transform(this[el.key])}</${el.key}>`;
        }
        return invoice_xml;
    }
}
exports.XmlElement = XmlElement;
class MappingElement {
    constructor(key, transform = null) {
        this.key = key;
        this.transform = transform || ((value) => value);
    }
}
exports.MappingElement = MappingElement;
