// @addo/addo-printer/lib/core/printer-error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrinterInfo = exports.PrinterError = void 0;
/**
* @example
* #define CP_PRINTERSTATUS_ERROR_CUTTER(error_status) (error_status & 0x01)
* #define CP_PRINTERSTATUS_ERROR_FLASH(error_status) (error_status & 0x02)
* #define CP_PRINTERSTATUS_ERROR_NOPAPER(error_status) (error_status & 0x04)
* #define CP_PRINTERSTATUS_ERROR_VOLTAGE(error_status) (error_status & 0x08)
* #define CP_PRINTERSTATUS_ERROR_MARKER(error_status) (error_status & 0x10)
* #define CP_PRINTERSTATUS_ERROR_ENGINE(error_status) (error_status & 0x20)
* #define CP_PRINTERSTATUS_ERROR_OVERHEAT(error_status) (error_status & 0x40)
* #define CP_PRINTERSTATUS_ERROR_COVERUP(error_status) (error_status & 0x80)
* #define CP_PRINTERSTATUS_ERROR_MOTOR(error_status) (error_status & 0x100)
*/
var PrinterError;
(function (PrinterError) {
    PrinterError[PrinterError["CUTTER"] = 1] = "CUTTER";
    PrinterError[PrinterError["FLASH"] = 2] = "FLASH";
    PrinterError[PrinterError["NOPAPER"] = 4] = "NOPAPER";
    PrinterError[PrinterError["VOLTAGE"] = 8] = "VOLTAGE";
    PrinterError[PrinterError["MARKER"] = 16] = "MARKER";
    PrinterError[PrinterError["ENGINE"] = 32] = "ENGINE";
    PrinterError[PrinterError["OVERHEAT"] = 64] = "OVERHEAT";
    PrinterError[PrinterError["COVERUP"] = 128] = "COVERUP";
    PrinterError[PrinterError["MOTOR"] = 256] = "MOTOR";
})(PrinterError = exports.PrinterError || (exports.PrinterError = {}));
/**
* @example
* #define CP_PRINTERSTATUS_INFO_LABELPAPER(info_status) (info_status & 0x02)
* #define CP_PRINTERSTATUS_INFO_LABELMODE(info_status) (info_status & 0x04)
* #define CP_PRINTERSTATUS_INFO_HAVEDATA(info_status) (info_status & 0x08)
* #define CP_PRINTERSTATUS_INFO_NOPAPERCANCELED(info_status) (info_status & 0x10)
* #define CP_PRINTERSTATUS_INFO_PAPERNOFETCH(info_status) (info_status & 0x20)
* #define CP_PRINTERSTATUS_INFO_PRINTIDLE(info_status) (info_status & 0x40)
* #define CP_PRINTERSTATUS_INFO_RECVIDLE(info_status) (info_status & 0x80)
*  */
var PrinterInfo;
(function (PrinterInfo) {
    PrinterInfo[PrinterInfo["LABELPAPER"] = 2] = "LABELPAPER";
    PrinterInfo[PrinterInfo["LABELMODE"] = 4] = "LABELMODE";
    PrinterInfo[PrinterInfo["HAVEDATA"] = 8] = "HAVEDATA";
    PrinterInfo[PrinterInfo["NOPAPERCANCELED"] = 16] = "NOPAPERCANCELED";
    PrinterInfo[PrinterInfo["INFO_PAPERNOFETCH"] = 32] = "INFO_PAPERNOFETCH";
    PrinterInfo[PrinterInfo["PRINTIDLE"] = 64] = "PRINTIDLE";
    PrinterInfo[PrinterInfo["RECVIDLE"] = 128] = "RECVIDLE";
})(PrinterInfo = exports.PrinterInfo || (exports.PrinterInfo = {}));
