// @addo/common-api/lib/src/utils/http/servers/nestjs/nestjs.microservice.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NestJsMicroservice = void 0;
const microservice_1 = require("../../microservice");
const http_request_1 = require("../../models/http.request");
class NestJsMicroservice extends microservice_1.Microservice {
    constructor(handler) {
        super(handler);
    }
    parseRequest(request) {
        let req = new http_request_1.Request();
        req.queryStringParameters = request.query;
        req.pathParameters = request.params;
        req.headers = JSON.parse(JSON.stringify(request.headers));
        req.body = request.body;
        req.original = request;
        return new http_request_1.HttpRequest(req);
    }
    async getAHandler(event, context, callback) {
        this.context = context;
        this.callback = callback;
        context.callbackWaitsForEmptyEventLoop = false;
        this.http = this.parseRequest(event);
        return await this.handle(this.http);
    }
}
exports.NestJsMicroservice = NestJsMicroservice;
